### ISSUE TITLE
Transfer progress notifications are prohibitively slow

### ISSUE THREAD

**Issue by bpeabody** [nodegit/nodegit#919]
Having even an empty transfer progress function for fetches adds a large amount of overhead. In the code below, for example, on the nodegit repo itself, specifying a transfer progress function adds about 38s for me.

The progress function is called once for each object fetched from the repository: about 25,000 times for the nodegit repo as of today. That would indicate an overhead of about 1.5ms for calling from C++ into JavaScript (running on a 3.1ghz i7). This amount of overhead feels a little high; I wonder if there might be some room for optimizing the JS/C++ binding/bridge code.

Even if significantly optimized (e.g., by a factor of 10), the overhead of calling from C++ into JS is probably high enough to be prohibitive; it might be good to be able to provide some conflation so that the progress function is called no more than a specified frequency (e.g., no more than once every 100ms).

If no one has an easy answer for this, I would be more than happy to lend a hand (I have a lot of experience with v8).

Sample code:

    const NodeGit = require("nodegit");

    console.time("no-progress");

    NodeGit.Clone("https://github.com/nodegit/nodegit.git", "no-progress")
    .then(function () {
        console.timeEnd("no-progress");
        console.time("progress");
        var options = new NodeGit.CloneOptions();
        var fetchOpts = new NodeGit.FetchOptions();
        options.fetchOpts = fetchOpts;
        var callbacks = new NodeGit.RemoteCallbacks();
        fetchOpts.callbacks = callbacks;
        callbacks.transferProgress = function () {
        };
        return NodeGit.Clone("https://github.com/nodegit/nodegit.git",
                             "progress",
                             options);
    }).done(function () {
        console.timeEnd("progress");
        process.exit(0);
    });

**Comment by johnhaley81** [comment]
Wow, that's huge. You're totally right we need to throttle this at the C level. 

The code that would handle this is in [callback_helpers.cc](https://github.com/nodegit/nodegit/blob/master/generate/templates/partials/callback_helpers.cc#L4-L31) and [field_accessors.cc](https://github.com/nodegit/nodegit/blob/master/generate/templates/partials/field_accessors.cc#L85-L112).

Those are the static methods that we send to libgit2 to use as the proxy for the JS callbacks that will eventually be called. If we throttle these then this will work.

We should also add some flags to the [callbacks](https://github.com/nodegit/nodegit/blob/master/generate/input/callbacks.json) to let the generator methods know that something should be throttled or not. This is probably something that we _could_ apply universally but it seems like we shouldn't. The only thing that should be getting spammed with libgit2 callbacks should be `progress_cb` functions.

If you want to take this up @bpeabody I will buy you a beer!

/cc @implausible

**Comment by bpeabody** [comment]
I agree that applying conflation across-the-board might not be a good idea. We can preserve the contract of the `progress_cb` functions through conflation, but I can imagine cases where the contract would be violated if the callback wasn't invoked each time... unless you could somehow marshall all the arguments for all the callbacks across to JS in one array and then invoke the callback multiple times... but that seems pretty complicated.

Let me see if I can take a whack at it. I would gladly accept a beer (delivery in NYC), but I am deeply in debt to this project already; you guys have done a great job.

**Comment by johnhaley81** [comment]
Implemented in #932

**Comment by tommoor** [comment]
Is this fixed up now?

**Comment by a-teammate** [comment]
seems so https://github.com/nodegit/nodegit/pull/944#issuecomment-207663296

close?
